### DISCUSSION TITLE
Now upsert is not persist/flush friendly

### DISCUSSION THREAD

**Discussion by darky** [mikro-orm/mikro-orm#4839] [General]
**Is your feature request related to a problem? Please describe.**
When I using `insert`, `update` or `delete` operations, then I happy to implement it via `persist`/`flush` flow.
But `upsert` doesn't support `persist`/`flush` flow, it's immediately run SQL query outside of transaction.
As workaround I can use `fetch`, then `insert` or `update` operations, which compatible with `persist`/`flush` flow,
but it's often inconvinient  and excess comparing to `upsert`

**Describe the solution you'd like**
Would be great, if `em.upsert` works in compatible way with `persist`/`flush` flow, but seems it's breaking change.
Maybe `em.upsertPersist` would be good for backward compatibility.

**Comment by B4nan** [comment]
Yes, the point of upsert is to run the query right ahead. It does respect the current EM state, if you try to upsert an entity that is already loaded to the context, it will just assign new data to its state and delay the update query for the flush call. After the upsert call resolves, the returned entity is part of your identity map and you can modify it in any way you want (and flush afterwards).

I am not really interested in mixing flush operation with upsert queries, flushing is about doing things we know how they should be done - we know what to insert and what to update. That way we can be sure what is the right order of the queries, with upsert we can't really do that. More importantly, without the upsert call actually doing things right ahead, there is no way to know what the actual state of the database is - the upsert call returns entities because it checks the state for you. I don't think this can actually work the way you want.

> As workaround I can use fetch, then insert or update operations, which compatible with persist/flush flow,

Fetch? there is no fetch method in the ORM.

**Reply by darky** [reply]
> if you try to upsert an entity that is already loaded to the context, it will just assign new data to its state and delay the update query for the flush call. After the upsert call resolves, the returned entity is part of your identity map and you can modify it in any way you want (and flush afterwards).

This implicit behaviour is confusing. As programmer, I don't want to think about this complex logic behind `upsert`. If I want this behaviour exactly ☝️ , I will do it in explicit way with populating of **Identity Map** via `findOne` or similar and then change entity (or create it, if not exists) for further `persist`/`flush`

> flushing is about doing things we know how they should be done - we know what to insert and what to update. That way we can be sure what is the right order of the queries, with upsert we can't really do that

I know how upsert should be done. Very often, I have **Entity** with few amount of columns and I want just `upsert`. Current state of DB not interesting for me. Just do `insert`, if not exists or do `update`, if exists. As programmer, I want to write one line of code with `upsert` (and be compatible with `persist`/`flow` of course), not multiple lines of code with boring `findOne`, then checking: if exists, then `update` or if not exists, then `insert`.
 About queries order: don't understand what's is problem. We can use `insert on confict do update` query or multiple queries: `update where` and then `insert on conflict do nothing`
 
 > Fetch? there is no fetch method in the ORM.
 
 I mean fetch in common sense. Ok, `findOne` for example.

**Reply by kingdun3284** [reply]
I have encountered the same problem. I think it is fine to separate the flush logic with upsert, but it should be mentioned in document that the logic in subscriber will not be executed.
